man/mdeliver.1
.Dd February 4, 2021
.Dt MDELIVER 1
.Os
.Sh NAME
.Nm mdeliver
.Nd deliver messages or import mbox file
.Sh SYNOPSIS
.Nm
.Op Fl c
.Op Fl v
.Op Fl X Ar flags
.Ar dir
<
.Ar message
.Nm
.Fl M
.Op Fl c
.Op Fl v
.Op Fl X Ar flags
.Ar dir
<
.Ar mbox
.Sh DESCRIPTION
.Nm
adds a message given on the standard input
as a new message in the maildir
.Ar dir .
.Pp
When
.Fl M
is used,
.Nm
will regard the standard input as
an MBOXRD mailbox, split it on
.Dq Li "From "
and deliver each message,
decoding it according to the MBOXRD convention.
In this case,
.Nm
will set the mtime according to the value of
.Sq Li Date\&:
and the maildir flags according to the value of
.Sq Li Status\&:
or
.Sq Li X-Status\&: .
.Pp
The messages are delivered in a reliable way and use the default
.Xr umask 2 .
.Pp
Please note that no syntactical checks are performed
on the messages.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl M
Deliver each message of an mbox.
.It Fl c
Deliver messages into
.Pa cur/ ,
not
.Pa new/
(the default).
.It Fl v
Print each new message filename after delivery.
.It Fl X Ar flg
Override the flags of the new message file to be
.Ar flg .
.El
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Xr mexport 1 ,
.Xr mrefile 1 ,
.Xr maildir 5 ,
.Xr mbox 5
.Pp
.Lk http://www.digitalpreservation.gov/formats/fdd/fdd000385.shtml "MBOXRD Email Format"
.Pp
.Lk https://cr.yp.to/proto/maildir.html "Using maildir format"
.Sh AUTHORS
.An Leah Neukirchen Aq Mt leah@vuxu.org
.Sh LICENSE
.Nm
is in the public domain.
.Pp
To the extent possible under law,
the creator of this work
has waived all copyright and related or
neighboring rights to this work.
.Pp
.Lk http://creativecommons.org/publicdomain/zero/1.0/


NEWS.md
## 1.2

* Caution! The tools mdeliver and mexport were buggy in handling and
  generation of trailing empty lines in MBOX-RD.  Do not import
  mbox files generated by mexport >=1.2 with mdeliver <1.2 if you
  require verbatim message delivery.

## 1.1 (2021-01-14)

* mcom: allow tilde prefixed path for profile's outbox setting
* mcom: detect and report mmime errors
* add contrib/mmailto, a handler for mailto: links
* Bug fixes.

## 1.0 (2020-09-12)

* Caution! Backwards incompatible changes:
  * As a message name, `-` now refers to the message on the standard input,
    and not the previous message anymore.  Use `.-` to refer to the previous
    message in a short way.
    The tools will print a warning if you use `-` and the standard input
    comes from a TTY.
  * mpick: use the -F flag to read script files.
  * mpick: remove msglist support, use plain mmsg(7) arguments.
* Many mblaze tools now make use of pledge(2) on OpenBSD.
* add contrib/mcount, a tool to count mails
* mrep: use Reply-From configuration to find From header
* Many bug fixes.

## 0.7 (2020-05-15)

* All tools now follow symlinks to mails.
* mdirs: add -a to list all subfolders, ignoring Maildir++ convention.
* mcom: add preview alias for show command.
* mrep/mbnc: allow only one message as argument.
* maddr: add -d to only print display name.
* mthread: add -r to reverse top-level order.
* mlist: print number of matches when message selection is in place.
* mpick: many improvements.
* Many bug fixes.

## 0.6 (2020-01-18)

* mfwd: prefix Subject with "Fwd:".
* mscan: add dottime formatting.
* mlist: look at maildir/new too for messages.
* Many bug and portability fixes.

## 0.5.1 (2019-03-03)

* mdeliver: preserve mtime in mrefile
* mdirs: add -0 to separate folders by NUL characters
* Fixes for buffer-overflows, found by fuzzing.
* Fixes for memleaks.

## 0.5 (2019-02-09)

* New tool msearch to wrap several mail indexers.
* New zsh completion _mblaze.
* mnext/mprev were removed (you can call `mless +`/`mless .-`).
* The GnuPG tools in contrib/ now use gpg2.
* mshow exits with error if it could not extract all attachments
* mrep: add -noquote to disable quoting the message replied to
* mdeliver: keep permissions of messages
* mcom: aborting the editor is now more like delete than cancel
* mcom: add -send to send directly without editing
* mcom: check if mail is formatted sensibly
* mpick: new flag -v for statistics
* mscan: new flag -v for statistics
* magrep: add -h, which is like -p but doesn't print the file name
* mscan: prioritize displaying trashed mail over other markers
* mpick: fix off-by-one in expression parsing
* Many bug fixes.

## 0.4 (2018-08-15)

* New tool mrefile to move or copy messages.
* contrib/mp7m: add application/pkcs7-mime decoder
* mcom: allow setting arbitrary headers from command line
* mcom: add -body to prepopulate drafts
* mcom: mark drafts as seen after sending
* mcom: update flags after mrep/mbnc/mfwd
* mshow: add -B to decode MIME parameters in broken mails
* magrep: add -l flag
* Many bug fixes.

## 0.3.2 (2018-02-13)

* magrep: add *:REGEX to search in any header
* Fix of a buffer overflow in blaze822_multipart.
* Small bug fixes.
* Many documentation improvements by Larry Hynes.

## 0.3.1 (2018-01-30)

* mless: support $NO_COLOR
* mcolor: support $NO_COLOR
* blaze822.h: ensure PATH_MAX is defined
* Improved documentation.
* Many fixes for address parser.

## 0.3 (2018-01-12)

* New tool mflow to reformat format=flowed plain text mails.
* New tool mbnc to bounce mails (send original mail to someone else).
* mshow filters can output raw text now, e.g. for HTML rendering with colors.
* mrep can quote mail that doesn't have a plain text part.
* mcom runs mmime when deemed necessary.
* mhdr can extract MIME parameters.
* New contrib: mrecode
* New contrib: mraw
* mshow regards non-MIME mails as MIME mails with one part now.
* mshow -F to disable MIME filters.
* mpick supports negations now.
* msed can remove headers depending on their value.
* Improved UTF-8 parsing.
* Improved documentation.
* Numerous bug fixes and portability fixes.

## 0.2 (2017-07-17)

* New sequence syntax `m:+n` for `n` messages after message `m`.
* Threading shortcuts `=`, `_`, `^` for `.=`, `._`, `.^`.
* Sequence related errors are now reported.
* minc and mlist normalize slashes in paths.
* mfwd now generates conforming message/rfc822 parts.
* mthread can add optional folders (e.g. your outbox) to resolve message ids.
* mcom now adds Date: just before sending or cancelling the mail.
* VIOLATIONS.md documents how mblaze works with certain common mistakes.
* Full documentation revamp by Larry Hynes.
* Fix rare crash looking for mail body.
* Numerous small bug and portability fixes.

## 0.1 (2017-06-24)

* Initial release


mdeliver.c
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>

#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <search.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <unistd.h>

#include "blaze822.h"
#include "xpledge.h"

/*
design rationale:
- only MBOX-RD because it's the only reasonable lossless encoding, and decodes
  MBOX-O fine
- date from Date: since From lines are usually crap
- proper maildir delivery because it's not that hard
- no creation of maildirs, should be a separate tool
*/

static int cflag;
static int kflag;
static int Mflag;
static int vflag;
static char *Xflag;

char *targetdir;
long delivery;
int preserve_mtime;
int try_rename;

char host[64];
void
gethost() {
	gethostname(host, sizeof host);
	// termination not posix guaranteed
	host[sizeof host - 1] = 0;
	// replace / and : with -
	char *s;
	for (s = host; *s; s++)
		if (*s == '/' || *s == ':')
			*s = '-';
}

int
deliver(char *infilename)
{
	int outfd;
	FILE *infile;
	FILE *outfile;
	char dst[PATH_MAX];
	char tmp[PATH_MAX];
	char id[128];
	struct timeval tv;

	char *line = 0;
	size_t linelen = 0;

	if (infilename) {
		if (try_rename) {
			infile = 0;
		} else {
			infile = fopen(infilename, "r");
			if (!infile) {
				fprintf(stderr, "mrefile: %s: %s\n",
				    infilename, strerror(errno));
				return -1;
			}
		}
	} else {
		// mdeliver
		infile = stdin;
	}

	if (Mflag) {
		// skip to first "From " line
		while (1) {
			errno = 0;
			ssize_t rd = getdelim(&line, &linelen, '\n', infile);
			if (rd == -1) {
				if (errno == 0)
					errno = EINVAL;  // invalid mbox file
				goto fail;
			}

			if (strncmp("From ", line, 5) == 0)
				break;
		}
	}

	do {
		delivery++;
try_again:
		gettimeofday(&tv, 0);

		snprintf(id, sizeof id, "%ld.M%06ldP%ldQ%ld.%s",
		    (long)tv.tv_sec, (long)tv.tv_usec, (long)getpid(),
		    delivery, host);

		snprintf(tmp, sizeof tmp, "%s/tmp/%s", targetdir, id);

		if (try_rename) {
			snprintf(dst, sizeof dst, "%s/%s/%s:2,%s",
			    targetdir, cflag ? "cur" : "new", id, Xflag);
			if (rename(infilename, dst) == 0)
				goto success;

			/* rename failed, open file and try copying */

			infile = fopen(infilename, "r");
			if (!infile) {
				fprintf(stderr, "mrefile: %s: %s\n",
				    infilename, strerror(errno));
				return -1;
			}
		}

		struct stat st;
		if (fstat(fileno(infile), &st) < 0)
			st.st_mode = 0600;
		if (S_ISFIFO(st.st_mode))
			st.st_mode = 0600;
		outfd = open(tmp, O_CREAT | O_WRONLY | O_EXCL,
		    st.st_mode & 07777);
		if (outfd < 0) {
			if (errno == EEXIST)
				goto try_again;
			if (errno == ENOENT)
				fprintf(stderr, "mrefile: %s/tmp: %s\n",
				    targetdir, strerror(errno));
			else
				fprintf(stderr, "mrefile: %s: %s\n",
				    tmp, strerror(errno));
			goto fail;
		}

		outfile = fdopen(outfd, "w");

		char statusflags[5] = { 0 };

		int in_header = 1;
		int is_old = 0;
		int prev_line_empty = 0;
		int this_line_empty = 0;  // only for mbox parsing
		while (1) {
			errno = 0;
			ssize_t rd = getdelim(&line, &linelen, '\n', infile);
			if (rd == -1) {
				if (errno != 0)
					goto fail;
				break;
			}
			char *line_start = line;

			if (line[0] == '\n' && (!line[1] ||
			                        (line[1] == '\r' && !line[2]))) {
				this_line_empty = Mflag ? 1 : 0;
				in_header = 0;
			} else {
				this_line_empty = 0;
			}

			if (Mflag && strncmp("From ", line, 5) == 0)
				break;

			if (Mflag && in_header &&
			    (strncasecmp("status:", line, 7) == 0 ||
			     strncasecmp("x-status:", line, 9) == 0)) {
				char *v = strchr(line, ':');
				if (v) {
					if (strchr(v, 'F')) statusflags[0] = 'F';
					if (strchr(v, 'A')) statusflags[1] = 'R';
					if (strchr(v, 'R')) statusflags[2] = 'S';
					if (strchr(v, 'D')) statusflags[3] = 'T';
					if (strchr(v, 'O')) is_old = 1;
				}
				continue;  // drop header
			}

			if (Mflag) {
				// MBOXRD: strip first > from >>..>>From
				char *s = line;
				while (*s == '>')
					s++;
				if (strncmp("From ", s, 5) == 0) {
					line_start++;
					rd--;
				}
			}

			// print delayed empty line
			if (prev_line_empty)
				if (fputc('\n', outfile) == EOF)
					goto fail;
			if (!this_line_empty)
				if (fwrite(line_start, 1, rd, outfile) != (size_t)rd)
					goto fail;

			prev_line_empty = this_line_empty;
		}
		if (fflush(outfile) == EOF)
			goto fail;
		if (fsync(outfd) < 0)
			goto fail;
		if (fclose(outfile) == EOF)
			goto fail;

		// compress flags
		int i, j;
		for (i = sizeof statusflags - 1; i >= 0; i--)
			if (!statusflags[i])
				for (j = i+1; j < (int)sizeof statusflags; j++)
					statusflags[j-1] = statusflags[j];

		if (Mflag) {
			struct message *msg = blaze822_file(tmp);
			time_t date = -1;
			char *v;

			if (msg && (v = blaze822_hdr(msg, "date"))) {
				date = blaze822_date(v);
				if (date != -1) {
					const struct timeval times[2] = {
						{ tv.tv_sec, tv.tv_usec },
						{ date, 0 }
					};
					utimes(tmp, times);
				}
			}
			blaze822_free(msg);
		}

		if (preserve_mtime) {
#if defined(AT_FDCWD) && defined(UTIME_NOW) && defined(UTIME_OMIT)
			const struct timespec times[2] = {
				{ tv.tv_sec, tv.tv_usec * 1000L },
#if (defined(__APPLE__) && defined(__MACH__))
			        st.st_mtimespec
#else /* POSIX.1-2008 */
				st.st_mtim
#endif
			};
			utimensat(AT_FDCWD, tmp, times, 0);
#else
			const struct timeval times[2] = {
				tv,
				{ st.st_mtime, 0 }
			};
			utimes(tmp, times);
#endif
		}

		snprintf(dst, sizeof dst, "%s/%s/%s:2,%s",
		    targetdir, (cflag || is_old) ? "cur" : "new", id,
		    Xflag ? Xflag : statusflags);
		if (rename(tmp, dst) != 0)
			goto fail;

success:
		if (vflag)
			printf("%s\n", dst);
	} while (Mflag && !feof(infile));

	if (infile)
		fclose(infile);
	return 0;

fail:
	if (infile)
		fclose(infile);
	return -1;
}

void
refile(char *file)
{
	while (*file == ' ' || *file == '\t')
		file++;

	// keep flags
	char *flags = strstr(file, ":2,");
	if (flags)
		Xflag = flags + 3;
	else
		Xflag = "";

	if (deliver(file) < 0) {
		perror("mrefile");
		return;
	}

	if (!kflag && !try_rename)
		unlink(file);
}

int
main(int argc, char *argv[])
{
	if (strchr(argv[0], 'f')) {
		// mrefile(1)

		cflag = 1;  // use cur/
		preserve_mtime = 1;
		try_rename = 1;

		int c;
		while ((c = getopt(argc, argv, "kv")) != -1)
			switch (c) {
			case 'k': kflag = 1; try_rename = 0; break;
			case 'v': vflag = 1; break;
			default:
usage:
				fprintf(stderr,
				    "Usage: mrefile [-kv] [msgs...] maildir\n");
			exit(1);
		}

		if (argc == optind)
			goto usage;

		targetdir = argv[argc - 1];
		gethost();

		if (argc == optind + 1 && isatty(0))
			goto usage;
		else
			blaze822_loop(argc - 1 - optind, argv + optind, refile);

		return 0;
	}

	int c;
	while ((c = getopt(argc, argv, "cMvX:")) != -1)
		switch (c) {
		case 'c': cflag = 1; break;
		case 'M': Mflag = 1; break;
		case 'v': vflag = 1; break;
		case 'X': Xflag = optarg; break;
		default:
usage2:
			fprintf(stderr,
"Usage: mdeliver [-c] [-v] [-X flags] dir < message\n"
"       mdeliver -M [-c] [-v] [-X flags] dir < mbox\n"
				);
			exit(1);
		}

	if (argc != optind+1)
		goto usage2;

	xpledge("stdio rpath wpath cpath", "");

	targetdir = argv[optind];

	gethost();

	if (deliver(0) < 0) {
		perror("mdeliver");
		return 2;
	}

	return 0;
}
